datalad/local/foreach.py

```diff
@@ -0,0 +1,275 @@
+# emacs: -*- mode: python; py-indent-offset: 4; tab-width: 4; indent-tabs-mode: nil -*-
+# ex: set sts=4 ts=4 sw=4 noet:
+# ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+#
+#   See COPYING file distributed along with the datalad package for the
+#   copyright and license terms.
+#
+# ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+"""Plumbing command for running a command on each (sub)dataset"""
+
+__docformat__ = 'restructuredtext'
+
+
+import logging
+
+from argparse import REMAINDER
+from itertools import chain
+from tempfile import mkdtemp
+
+from datalad.cmd import NoCapture, StdOutErrCapture
+from datalad.core.local.run import normalize_command
+
+from datalad.distribution.dataset import (
+    Dataset,
+    EnsureDataset,
+    datasetmethod,
+    require_dataset,
+)
+from datalad.dochelpers import exc_str
+from datalad.interface.base import (
+    Interface,
+    build_doc,
+)
+from datalad.interface.common_opts import (
+    jobs_opt,
+    recursion_flag,
+    recursion_limit,
+)
+from datalad.interface.results import get_status_dict
+from datalad.interface.utils import eval_results
+from datalad.support.constraints import (
+    EnsureBool,
+    EnsureNone,
+)
+from datalad.support.parallel import (
+    ProducerConsumer,
+    ProducerConsumerProgressLog,
+    no_parentds_in_futures,
+    no_subds_in_futures,
+)
+from datalad.support.param import Parameter
+from datalad.utils import (
+    SequenceFormatter,
+    getpwd,
+    shortened_repr,
+)
+
+lgr = logging.getLogger('datalad.local.foreach')
+
+
+@build_doc
+class ForEach(Interface):
+    r"""Run a command on the dataset and/or each of its sub-datasets.
+
+    Each dataset's repo runner is reused to execute the commands, so they
+    are executed in the top directory of a corresponding dataset.
+    In contrast, Python commands (expressions) are evaluated without changing directory.
+    TODO: unify via explicit --chdir? but not sure if worth it
+
+    WARNING: Python expressions are `eval`ed within the scope of this process
+    and provided existing objects.
```

Comment:
Update: I now saw that `eval()` is called with `globals`, so I don't really understand the warning anymore.

Reply:
unlike with external execution, I thought you could potentially alter the provided instance (`ds`) or just inflict some other change which would persist in the same process (e.g. whenever `foreach` used through Python API)...  I guess situation is not much different from having a command doing changes or evil externally but still feels that it is worth noting somehow that  we are just evaluating in the same process.  I guess it could be `FYI:` or just `Note:` instead of a `WARNING:`

Reply:
Decided not to bother with any notes etc.